connection: Add support for socket options on created client connections

jrajahalme opened the pull request:
Add an Options visitor to Network::ConnectionSocket. Listener filters
can set the visitor for any newly accepted connections, causing the
visitor is called in two sitations:

bool setOptions(ConnectionSocket&): This visitor method is called
right after a client socket is created for forwarding traffic received
from an accepted connection, before any traffic (including connection
setup) is sent on the new socket. This can be used to set socket
options, for example.

uint32_t hashKey(): This visitor is called on upstream connection pool
lookup. This gives the visitor the opportunity to separate upstream
connection pools so that connections for which a different values are
returned will not share connection pools. The purpose of this is to
allow the listener filter to make sure that a reused connection has
the required options as set by setOptions(). If this segmentation is
not needed the implementation should return zero.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

mattklein123 commented:
@jrajahalme thanks for the contribution. To kick off the discussion at a design level, I will admit that I'm not super thrilled about plumbing through "so_mark" everywhere. It's a very Linux (and iptables no less) concept. I'm also generally concerned that there will be other things like this in the future.

I would like to propose a general ID which could be loosely described as a "ConnectionModifier." It could be a visitor pattern class that takes connections and can modify them as needed. I'm wondering if we could hide all of the SO_MARK functionality inside such a class and pass this around. This has two benefits:

1) If we need to add something else later like SO_MARK we can easily do it.
2) It will hide very Linux specific functionality mostly behind an interface.

Thoughts?

jrajahalme commented:
@mattklein123 Not hung up on the particular implementation at all and I like your thought on extensibility. Do we have examples of the visitor pattern somewhere in the code base I could model this after, or could you provide an useful external reference so that we minimize the churn on this PR?

mattklein123 commented:
@jrajahalme 

I'm thinking something roughly like the following (I haven't really thought this fully through):

```
class ConnectionSocketModifier {
public:
  virtual modify(ConnectionSocket& socket) PURE;
}
```

Perhaps a listener filter could provide a `ConnectionSocketModifier` for further use by other parts of the system. For example, in your case, you want to provide a modifier which is used for connections created when passed a load balancer context for a certain downstream connection. So perhaps the flow looks like:

1) Filter attaches a modifier to the downstream connection.
2) When a new upstream connection is created it optionally passes the downstream modifier in.

I think doing this we can almost entirely hide SO_MARK from Envoy (in fact you could just have it in your private code in whatever you are doing). The one gotcha here is the part of your change where you want to segment the connection pools by the mark. I think we should develop some other generic way of doing this, potentially driven by the generic modifier. E.g., return a generic index or something.

mattklein123 commented:
cc @alyssawilk and @ggreenway for any comments on the above discussion.

alyssawilk commented:
SGTM.  My extensibility check is often "will this work for upcoming features I know of" and I think this proposal would work for both.  
  - setting QoS/ToS both on connection creation and "in flight" edits based on DoS / response type / backend headers / etc
 - QUIC, when you have to set options on the virtual socket (latch them in the in-userspace quic connection then set as per-packet options on the mega UDP fd) rather than setting them on an fd but the quic connection can then implement this interface

jrajahalme commented:
I started on a revision before seeing the comments today, and took the liberty to rebase and force push a new proposal on this PR. It implements what @mattklein123 suggests above, just the names of identifiers are different. I managed to not mention "mark" once in this revision, and the tests should now work on all operating systems.

I have a follow-up PR that allows options to be set on the listener socket itself, but wanted to have this through first to limit the scope of the discussion at this time.

mattklein123 commented:
@jrajahalme at quick glance this approach looks great to me. I would give it a more thorough look tomorrow. Thanks!

mattklein123 commented on include/envoy/network/connection.h:
```diff
@@ -216,6 +220,13 @@ class Connection : public Event::DeferredDeletable, public FilterManager {
+   * traffic received on this connection.
+   */
+  virtual const ConnectionSocket::OptionsSharedPtr&
+  socketOptionsForUpstreamConnections() const PURE;
```
From an interface perspective this bothers me a bit. There really shouldn't any logical association between a connection and an "upstream connection." Is there any other way we can express this? I'm wondering if it's as simple as actually just calling this `socketOptions()` to express the options for this connection, which would be used for upstream connections or any other purpose by other parts of the system. Thoughts?

jrajahalme replied:
On a second thought I think this is possible if we make the distinction of the calling site/purpose on the visitor name, i.e., we inherit the socketOptions from server to the associated client connections and the client connections then call the visitor as "setClientSocketOptions()" etc. This better?

mattklein123 replied:
Marginally (I will take it over what we have now), but not a ton. In the case of marking, aren't you basically trying to transfer options? Why can't we just expose options from a connection and then in other code use those options to match a new connection to an old one?

jrajahalme replied:
Let me see if I understand this right.. These calls should be like (naming is elaborate to make the point):

setSocketOptionsForNewConnections() - set the options to be exposed
socketOptionsForNewConnections() - get the exposed options, if any.

and then the callback/visitor to apply the options to the new connection:

if (options) {
  options->setOptions(ConnectionSocket&);
}

I.e., there is "by definition" no confusion about the options being applied to the connection they are set at, because they are only ever to be transferred to new connections?

So this is just about naming the members and how to document them? I'm fine naming these members as you proposed in your first comment above, I'll try to make it clear in the comments that these options do not apply to the connection on which these calls are made, but only to any new connections they are transferred to.

mattklein123 replied:
My main objection is that a connection is a connection. Tying anything to a connection about new connections is pretty fragile and IMO a poor abstraction. I will freely admit that this kind of conversation makes me vomit a little on the inside, but this is a bit of a stretch even for me. 😉 

So as I see it there are 2 options:
1) Have a connection expose its options, with a visitor which allows new connections to duplicate those options. It should not having anything to do with upstream, etc. Does this work for you?
2) We need to figure out some way to express options that should be used "upstream" in some other way, which means probably going outside of this interface. 

(1) is easier if that works (and I think it might?). (2) will require larger interface changes elsewhere.

alyssawilk replied:
Coming in late, +1 to this.  We'll probably eventually have option we want to apply to the downstream socket as well, so keeping the API as neutrally named as possible while making sure the right options go to the right places would be ideal.

mattklein123 commented on include/envoy/network/listen_socket.h:
```diff
@@ -90,6 +90,39 @@ class ConnectionSocket {
+     * @return bits that can be used to separate upstream connections based on the options. Should
+     *         return zero if connections with different options can be pooled together.
+     */
+    virtual uint32_t hashKey() const PURE;
```
At the interface level can we make this uint64_t? Seems like it should just be 64-bit. If the implementation chooses to just use the low 32-bits that's an implementation detail?

jrajahalme replied:
I was going back and forth with this a bit. The problem with dropping the upper bits is that then wrong connections could be reused. For example, let's say based on the server connections we want to apply options identified with '0x1234' to upstream connections. For another server connection we want to apply options identified with '0xff34' to the upstream connections. If the implementation drops the upper half and only considers '0x34' then it would reuse the first connection in the second case, and the traffic would be forwarded with e.g., wrong socket mark.

Maybe I should not have used the word "hash" in this context. It came from the fact that we use the return value as part of a connection pool hash key.

The cleanest interface I can think of would take in a std::vector<uint8_t> and the visitor can then append as many bytes as it needs to the hash key. Internally we would init it with one byte covering the Http::Protocol and priority. Better?

mattklein123 replied:
It seems like the user controls both sides, right? So if we do 64-bit, wouldn't the top 32-bits always be zero? I'm confused what the problem is.

jrajahalme replied:
The other side is in cluster_manager_impl, where we pick a connection pool from the connection pool container. Similar to the HTTP/1 vs. HTTP/2 or priority, we have to keep upstream connections with different options separate, i.e., keep them in separate pools. This piece is not pluggable, so it is not controlled by the filter setting the options.

To achieve the connection separation, I've changed the connection pool container from an array to an unordered_map. For this to work the options (or an encoding of them, depending on what the options are) need to be in the map key.

Right now in the PR I've assembled a uint64_t key from 1 bit for the priority, 2 bits  for the Http::Protocol and 32 bits from here. This works for my intended use case (socket mark), and I see that more bits would be needed if there were some QoS options, ToS bits etc. in the play. But even then, uint64_t for this purpose might be too restrictive.

One version I had locally had this as uint64_t, but with a comment that the top three bits may be ignored by the caller. But imagine if we add more protocols, etc, and have to change it to "top 4 bits being ignored", existing filter implementations could be broken.

See the problem now?

mattklein123 replied:
Yup fair enough. 32-bit seems fine. I would add more comments explaining why this is the case.

mattklein123 commented on include/envoy/network/listen_socket.h:
```diff
@@ -90,6 +90,39 @@ class ConnectionSocket {
+   * Set the socket options to be set on all upstream ClientConnections created to forward
+   * traffic received on this socket.
+   */
+  virtual void setOptionsForUpstreamConnections(const OptionsSharedPtr&) PURE;
```
Likewise would prefer to remove the concept of upstream here. Thoughts on removing it?

jrajahalme replied:
While addressing the comment above I realized that I had not applied the visitor pattern to it's full potential here, see the other response.

mattklein123 commented on include/envoy/network/connection.h:
```diff
@@ -85,6 +86,9 @@ class Connection : public Event::DeferredDeletable, public FilterManager {
     Stats::Counter* bind_errors_;
+    // Counter* as this is an optional counter. Socket options errors will not be tracked if this is
+    // nullptr.
+    Stats::Counter* socket_options_errors_;
```
Unless you actually wire this up in production code I would kill it and associated logic. I don't see it wired up anywhere but maybe I missed it.

jrajahalme replied:
Agree, this is a remnant from the initial straightforward socket mark version. Now the visitor can observe the success/failure of the setsockopt directly, no need for this.

mattklein123 commented:
>All tests seem to pass with this change and the close reason is local rather than remote if the connect() call fails, so maybe this is ok.

Unfortunately I think this is probably not OK, since it will effect stats. I think a connect error should count as remote close. Can you add a test which checks this and would have broken with this change?

jrajahalme commented:
> On Jan 29, 2018, at 2:31 PM, Matt Klein <notifications@github.com> wrote:
> 
> All tests seem to pass with this change and the close reason is local rather than remote if the connect() call fails, so maybe this is ok.
> 
> Unfortunately I think this is probably not OK, since it will effect stats. I think a connect error should count as remote close. Can you add a test which checks this and would have broken with this change?
> 
> 

Will do.

jrajahalme commented:
> On Jan 29, 2018, at 2:31 PM, Matt Klein <notifications@github.com> wrote:
> 
> All tests seem to pass with this change and the close reason is local rather than remote if the connect() call fails, so maybe this is ok.
> 
> Unfortunately I think this is probably not OK, since it will effect stats. I think a connect error should count as remote close. Can you add a test which checks this and would have broken with this change?
> 
> 

Which stats are affected?

mattklein123 commented:
There are quite a few stats which are driven by local/remote close to determine which side terminated a connection.

lizan commented on include/envoy/event/dispatcher.h:
```diff
@@ -52,12 +52,15 @@ class Dispatcher {
                          Network::Address::InstanceConstSharedPtr source_address,
-                         Network::TransportSocketPtr&& transport_socket) PURE;
+                         Network::TransportSocketPtr&& transport_socket,
+                         const Network::ConnectionSocket::OptionsSharedPtr& options = nullptr) PURE;
```
Default arguments for virtual functions are not allowed:
https://google.github.io/styleguide/cppguide.html#Default_Arguments

mattklein123 replied:
+1, thanks for catching.

jrajahalme replied:
Thanks for pointing this out, fixed.

lizan commented on source/common/upstream/cluster_manager_impl.h:
```diff
@@ -197,13 +197,12 @@ class ClusterManagerImpl : public ClusterManager, Logger::Loggable<Logger::Id::u
+      uint64_t key(ResourcePriority priority, Http::Protocol protocol, uint32_t hash_key) {
         ASSERT(NumResourcePriorities == 2); // One bit needed for priority
-        return enumToInt(protocol) << 1 | enumToInt(priority);
+        ASSERT(Http::NumProtocols <= 4);    // Two bits needed for priority
```
Can this be `static_assert`?

jrajahalme replied:
Yes, changed.

lizan commented on test/common/network/connection_impl_test.cc:
```diff
@@ -210,6 +210,132 @@ TEST_P(ConnectionImplTest, CloseDuringConnectCallback) {
+
+  auto options = std::make_shared<MockSocketOptions>();
+
+  EXPECT_CALL(*options, setOptions(_)).WillOnce(Invoke([&](Network::ConnectionSocket&) -> bool {
```
`.WillOnce(Return(false))`

jrajahalme replied:
OK.

lizan commented on test/common/network/connection_impl_test.cc:
```diff
@@ -210,6 +210,132 @@ TEST_P(ConnectionImplTest, CloseDuringConnectCallback) {
+
+  auto options = std::make_shared<MockSocketOptions>();
+
+  EXPECT_CALL(*options, setOptions(_)).WillOnce(Invoke([&](Network::ConnectionSocket&) -> bool {
```
`.WillOnce(Return(true))`

jrajahalme replied:
OK.

jrajahalme commented:
I added the test for immediate connect failures, then found out file events don't trigger on Mac in that case and added an explicit event trigger. Seems to work now.

mattklein123 commented on include/envoy/network/listen_socket.h:
```diff
@@ -90,6 +90,39 @@ class ConnectionSocket {
+    virtual bool setOptions(ConnectionSocket& socket) const PURE;
+
+    /**
+     * @return bits that can be used to separate upstream connections based on the options. Should
```
nit: remove "upstream". The concept applies to any connection I think.

jrajahalme replied:
Removed.

mattklein123 commented on source/common/http/http1/conn_pool.h:
```diff
@@ -139,8 +141,9 @@ class ConnPoolImpl : Logger::Loggable<Logger::Id::pool>, public ConnectionPool::
-                   Upstream::ResourcePriority priority)
-      : ConnPoolImpl(dispatcher, host, priority) {}
+                   Upstream::ResourcePriority priority,
+                   const Network::ConnectionSocket::OptionsSharedPtr& options = nullptr)
```
nit: default param here seems not worth it.

jrajahalme replied:
I thought I had all of these removed, checked and this was the last one. Thanks!

mattklein123 commented on source/common/network/connection_impl.cc:
```diff
@@ -477,22 +475,33 @@ void ConnectionImpl::updateWriteBufferStats(uint64_t num_written, uint64_t new_s
+  if (options) {
+    if (!options->setOptions(*socket_)) {
+      // Set a special error state to ensure asynchronous close to give the owner of the
+      // ConnectionImpl a chance to add callbacks and detect the "disconnect"
```
nit: period end of sentence.

jrajahalme replied:
Fixed.

mattklein123 commented on source/common/network/connection_impl.cc:
```diff
@@ -507,10 +516,12 @@ void ClientConnectionImpl::connect() {
       connecting_ = false;
       ENVOY_CONN_LOG(debug, "immediate connection error: {}", *this, errno);
+
+      // Trigger a write event. This is needed on Mac and seems harmless on Linux.
```
nit: s/Mac/OSX

jrajahalme replied:
Changed.

mattklein123 commented on source/common/network/listen_socket_impl.h:
```diff
@@ -78,9 +79,12 @@ class ConnectionSocketImpl : virtual public ConnectionSocket {
 
 protected:
   int fd_;
+  OptionsSharedPtr options_{nullptr};
```
nit: `{}` not needed.

jrajahalme replied:
Removed.

mattklein123 commented on source/common/upstream/cluster_manager_impl.cc:
```diff
@@ -604,26 +604,22 @@ void ClusterManagerImpl::ThreadLocalClusterManagerImpl::drainConnPools(
-    if (pool) {
-      container.drains_remaining_++;
-    }
+  std::vector<Http::ConnectionPool::Instance*> to_be_drained;
```
Why do we need to copy in temporary first?

jrajahalme replied:
We did when I initially made this change couple of months ago, but now I see there is the explicit break out in the case the container was nuked. I'll revert this and retest.

mattklein123 commented on source/common/upstream/cluster_manager_impl.cc:
```diff
@@ -604,26 +604,22 @@ void ClusterManagerImpl::ThreadLocalClusterManagerImpl::drainConnPools(
-    }
+  std::vector<Http::ConnectionPool::Instance*> to_be_drained;
+  for (const auto& pair : container.pools_) {
+    const Http::ConnectionPool::InstancePtr& pool = pair.second;
```
if this code ends up staying, local var not needed, same below.

jrajahalme replied:
IMO this aids in readability by making it visible what type pair.second is but I'll remove these if they stay.

mattklein123 commented on source/common/upstream/cluster_manager_impl.cc:
```diff
@@ -768,14 +761,31 @@ ClusterManagerImpl::ThreadLocalClusterManagerImpl::ClusterEntry::connPool(
   }
 
+  // Inherit socket options from downstream connection, if set.
+  uint32_t hash_key = 0;
```
Use `Optional<uint32_t>` to avoid maintaining 2 variables.

jrajahalme replied:
Changed.

ggreenway commented:
I took a quick look.  The interface for this lgtm.
